github/ci/services/common/k8s/pkg/wait/wait.go

```diff
@@ -0,0 +1,180 @@
+package wait
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strings"
+	"sync"
+	"time"
+
+	certmanagerv1 "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+	cmmeta "github.com/jetstack/cert-manager/pkg/apis/meta/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+	"kubevirt.io/project-infra/github/ci/services/common/k8s/pkg/client"
+)
+
+const poll = 5 * time.Second
+
+type stopChan struct {
+	c chan struct{}
+	sync.Once
+}
+
+func newStopChan() *stopChan {
+	return &stopChan{c: make(chan struct{})}
+}
+
+func (s *stopChan) closeOnce() {
+	s.Do(func() {
+		close(s.c)
+	})
+}
+
+func ForDeploymentReady(namespace, name string) {
+	clientset, err := client.NewClientset()
+	if err != nil {
+		log.Fatalf("Cloud not create clientset %v", err)
+	}
+
+	stop := newStopChan()
+
+	watchlist := cache.NewListWatchFromClient(clientset.AppsV1().RESTClient(), "deployments", namespace, fields.Everything())
+	_, controller := cache.NewInformer(watchlist, &appsv1.Deployment{}, poll, cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(o, n interface{}) {
+			newDeployment := n.(*appsv1.Deployment)
+
+			if newDeployment.Name != name {
+				return
+			}
+			if isDeploymentReady(newDeployment) {
+				stop.closeOnce()
+				return
+			}
+		},
+	})
+	controller.Run(stop.c)
+}
+
+func ForCertificateReady(namespace, name string) {
+	clientset, err := client.NewCertManagerClientset()
+	if err != nil {
+		log.Fatalf("Cloud not create clientset %v", err)
+	}
+
+	stop := newStopChan()
+
+	watchlist := cache.NewListWatchFromClient(clientset.CertmanagerV1().RESTClient(), "certificates", namespace, fields.Everything())
+	_, controller := cache.NewInformer(watchlist, &certmanagerv1.Certificate{}, poll, cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(o, n interface{}) {
+			newCertificate := n.(*certmanagerv1.Certificate)
+
+			if newCertificate.Name != name {
+				return
+			}
+			if isCertificateReady(newCertificate) {
```

Comment:
I think this logic is small enough, but just a heads up that no sophisticated time consuming logic mus exist here. The go scheduler has interesting side effects with locks and go routines if a go routine locks mutexes for a longer period.

Reply:
Yep, in this case these are just function extractions to improve readability, each `is*Ready` function only checks existing values in the object sent as parameter.